ddtelemetry/src/ipc/platform/unix/channel/async_channel.rs

```diff
@@ -0,0 +1,131 @@
+    net::UnixStream,
+};
+
+use super::{Channel, ChannelMetadata, PlatformHandle, MAX_FDS};
+
+#[derive(Debug)]
+#[pin_project]
+pub struct AsyncChannel {
+    #[pin]
+    inner: UnixStream,
+    pub metadata: Arc<Mutex<ChannelMetadata>>,
+}
+
+impl From<UnixStream> for AsyncChannel {
+    fn from(stream: UnixStream) -> Self {
+        AsyncChannel {
+            inner: stream,
+            metadata: Arc::new(Mutex::new(ChannelMetadata::default())),
+        }
+    }
+}
+
+impl TryFrom<Channel> for AsyncChannel {
+    type Error = io::Error;
+
+    fn try_from(value: Channel) -> Result<Self, Self::Error> {
+        let fd = PlatformHandle::<StdUnixStream>::from(value).into_instance()?;
+
+        fd.set_nonblocking(true)?;
+        Ok(AsyncChannel {
+            inner: UnixStream::from_std(fd)?,
+            metadata: Arc::new(Mutex::new(ChannelMetadata::default())),
+        })
+    }
+}
+
+impl AsyncWrite for AsyncChannel {
+    fn poll_write(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let project = self.project();
+        let handles = project.metadata.lock().unwrap().drain_to_send();
+
+        if !handles.is_empty() {
+            let fds: Vec<RawFd> = handles.iter().map(AsRawFd::as_raw_fd).collect();
+            match project.inner.send_with_fd(buf, &fds) {
+                Ok(sent) => {
+                    project
+                        .metadata
+                        .lock()
+                        .unwrap()
+                        .defer_close_handles(handles);
+                    Poll::Ready(Ok(sent))
+                }
+                Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                    project
+                        .metadata
+                        .lock()
+                        .unwrap()
+                        .reenqueue_for_sending(handles);
+                    project.inner.poll_write_ready(cx).map_ok(|_| 0)
+                }
+                Err(err) => Poll::Ready(Err(err)),
+            }
+        } else {
+            project.inner.poll_write(cx, buf)
+        }
+    }
+
+    fn poll_flush(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), io::Error>> {
+        self.project().inner.poll_flush(cx)
+    }
+
+    fn poll_shutdown(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), io::Error>> {
+        self.project().inner.poll_shutdown(cx)
+    }
+}
+
+impl AsyncRead for AsyncChannel {
+    fn poll_read(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let project = self.project();
+        let mut fds = [0; MAX_FDS];
+
+        unsafe {
+            let b = &mut *(buf.unfilled_mut() as *mut [std::mem::MaybeUninit<u8>] as *mut [u8]);
```

Comment:
Yeah, I think filling with zero's would be safer, and more secure. 
I basically followed what tokio AsyncRead was doing here. 

This is part of their optimization - since its a buffer that will be written to - then the contents of memory don't matter as much. But I'd prefer to play  on the safe side here